File: app/src/main/java/com/base/and/base/BaseTabFragment.java
package com.base.and.base;

import android.content.Context;
import android.os.Bundle;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.fragment.app.FragmentPagerAdapter;
import android.text.TextUtils;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.animation.AccelerateInterpolator;
import android.view.animation.DecelerateInterpolator;
import android.widget.LinearLayout;

import com.base.and.R;
import com.base.and.databinding.FragmentBaseTabBinding;
import com.base.and.utils.DensityUtil;
import com.base.and.view.ScaleTransitionPagerTitleView;

import net.lucode.hackware.magicindicator.ViewPagerHelper;
import net.lucode.hackware.magicindicator.buildins.commonnavigator.CommonNavigator;
import net.lucode.hackware.magicindicator.buildins.commonnavigator.abs.CommonNavigatorAdapter;
import net.lucode.hackware.magicindicator.buildins.commonnavigator.abs.IPagerIndicator;
import net.lucode.hackware.magicindicator.buildins.commonnavigator.abs.IPagerTitleView;
import net.lucode.hackware.magicindicator.buildins.commonnavigator.indicators.LinePagerIndicator;
import net.lucode.hackware.magicindicator.buildins.commonnavigator.titles.SimplePagerTitleView;

public abstract class BaseTabFragment extends BaseFragment<FragmentBaseTabBinding> {
    protected String[] mTabsName;

    @Override
    public int initContentView(LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        return R.layout.fragment_base_tab;
    }

    @Override
    public void initData() {
        if (TextUtils.isEmpty(getTitle())) {
            binding.titleLayout.rlRoot.setVisibility(View.GONE);
        } else {
            binding.titleLayout.title.setText(getTitle());
        }
        mTabsName = getTabNames();
        binding.viewPager.setAdapter(new FragmentPagerAdapter(getChildFragmentManager()) {
            @Override
            public Fragment getItem(int position) {
                return getTabItem(position);
            }

            @Override
            public int getCount() {
                return mTabsName.length;
            }

            @Override
            public CharSequence getPageTitle(int position) {
                return mTabsName[position];
            }
        });
        binding.viewPager.setOffscreenPageLimit(mTabsName.length);

        CommonNavigator commonNavigator = new CommonNavigator(getContext());
        commonNavigator.setAdjustMode(true);
        commonNavigator.setAdapter(new CommonNavigatorAdapter() {
            @Override
            public int getCount() {
                return mTabsName.length;
            }

            @Override
            public IPagerTitleView getTitleView(Context context, int index) {
                return getTabTitleView(context, index);
            }

            @Override
            public IPagerIndicator getIndicator(Context context) {
                LinePagerIndicator indicator = new LinePagerIndicator(context);
                indicator.setMode(LinePagerIndicator.MODE_EXACTLY);
                indicator.setLineHeight(DensityUtil.dp2px(context, 3));
                indicator.setLineWidth(DensityUtil.dp2px(context, 54));
                indicator.setRoundRadius(DensityUtil.dp2px(context, 2));
                indicator.setStartInterpolator(new AccelerateInterpolator());
                indicator.setEndInterpolator(new DecelerateInterpolator(2.0f));
                indicator.setColors(0xffEC1D24);
                return indicator;
            }

            @Override
            public float getTitleWeight(Context context, int index) {
                return getTabWeight(index);
            }
        });
        binding.tabs.setNavigator(commonNavigator);
//        LinearLayout titleContainer = commonNavigator.getTitleContainer();
//        titleContainer.setDividerPadding(200);
        ViewPagerHelper.bind(binding.tabs, binding.viewPager);
    }

    /**
     * tab的宽度权重 默认平分
     *
     * @param index
     * @return
     */
    protected float getTabWeight(int index) {
        return 1;
    }

    /**
     * titlebar标题 传空即不显示
     *
     * @return
     */
    protected abstract String getTitle();

    /**
     * tab标题数组
     *
     * @return
     */
    protected abstract String[] getTabNames();

    /**
     * tab对应的fragment
     *
     * @param position
     * @return
     */
    protected abstract Fragment getTabItem(int position);

    /**
     * 调用此方法设置分割线样式
     *
     * @param height
     * @param color
     */
    protected void setDivider(int height, int color) {
        LinearLayout.LayoutParams params = (LinearLayout.LayoutParams) binding.viewDivider.getLayoutParams();
        params.height = DensityUtil.dp2px(getContext(), height);
        binding.viewDivider.setLayoutParams(params);
        binding.viewDivider.setBackgroundColor(color);
        binding.viewDivider.setVisibility(View.VISIBLE);
    }

    /**
     * 可以复用此方法自定义tab的titleview样式
     *
     * @param context
     * @param index
     * @return
     */
    protected IPagerTitleView getTabTitleView(Context context, final int index) {
        SimplePagerTitleView simplePagerTitleView = new ScaleTransitionPagerTitleView(context);
        simplePagerTitleView.setText(mTabsName[index]);
        simplePagerTitleView.setTextSize(16);
        simplePagerTitleView.setNormalColor(0xff666666);
        simplePagerTitleView.setSelectedColor(0xff323232);
        simplePagerTitleView.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                binding.viewPager.setCurrentItem(index);
            }
        });
        return simplePagerTitleView;
    }
}

File: app/src/main/kotlin/com/jeahwan/origin/base/BaseTabFragment.kt
package com.jeahwan.origin.base

import android.content.Context
import android.text.TextUtils
import android.view.View
import android.view.ViewGroup
import android.view.animation.AccelerateInterpolator
import android.view.animation.DecelerateInterpolator
import android.widget.LinearLayout
import androidx.fragment.app.Fragment
import androidx.fragment.app.FragmentPagerAdapter
import com.jeahwan.origin.R
import com.jeahwan.origin.databinding.FragmentBaseTabBinding
import com.jeahwan.origin.utils.DensityUtil
import com.jeahwan.origin.view.ScaleTransitionPagerTitleView
import net.lucode.hackware.magicindicator.ViewPagerHelper
import net.lucode.hackware.magicindicator.buildins.commonnavigator.CommonNavigator
import net.lucode.hackware.magicindicator.buildins.commonnavigator.abs.CommonNavigatorAdapter
import net.lucode.hackware.magicindicator.buildins.commonnavigator.abs.IPagerIndicator
import net.lucode.hackware.magicindicator.buildins.commonnavigator.abs.IPagerTitleView
import net.lucode.hackware.magicindicator.buildins.commonnavigator.indicators.LinePagerIndicator
import net.lucode.hackware.magicindicator.buildins.commonnavigator.titles.SimplePagerTitleView

abstract class BaseTabFragment : BaseFragment<FragmentBaseTabBinding>(R.layout.fragment_base_tab) {

    protected lateinit var mTabsName: Array<String>

    override fun initData() {
        if (TextUtils.isEmpty(title)) {
            binding.titleLayout.rlRoot.visibility = View.GONE
        } else {
            binding.titleLayout.title = title
        }
        mTabsName = getTabNames()
        binding.viewDivider.visibility = if (showTabDivider()) View.VISIBLE else View.GONE
        binding.viewPager.adapter = object :
            FragmentPagerAdapter(childFragmentManager, BEHAVIOR_RESUME_ONLY_CURRENT_FRAGMENT) {
            override fun getItem(position: Int): Fragment {
                return getTabItem(position)
            }

            override fun getCount(): Int {
                return mTabsName.size
            }

            override fun getPageTitle(position: Int): CharSequence {
                return mTabsName[position]
            }
        }
        binding.viewPager.offscreenPageLimit = mTabsName.size
        if (!isAdjustMode) {
            val param = binding.tabs.layoutParams
            param.width = ViewGroup.LayoutParams.WRAP_CONTENT
            binding.tabs.layoutParams = param
        }
        val commonNavigator = CommonNavigator(context)
        commonNavigator.isAdjustMode = isAdjustMode
        commonNavigator.adapter = object : CommonNavigatorAdapter() {
            override fun getCount(): Int {
                return mTabsName.size
            }

            override fun getTitleView(context: Context, index: Int): IPagerTitleView {
                return getTabTitleView(context, index)
            }

            override fun getIndicator(context: Context): IPagerIndicator {
                return getTabIndicator(context)
            }

            override fun getTitleWeight(context: Context, index: Int): Float {
                return getTabWeight(index)
            }
        }
        binding.tabs.navigator = commonNavigator
        //        LinearLayout titleContainer = commonNavigator.getTitleContainer();
//        titleContainer.setDividerPadding(200);
        ViewPagerHelper.bind(binding.tabs, binding.viewPager)
    }

    /**
     * tab的宽度权重 默认平分
     *
     * @param index
     * @return
     */
    protected fun getTabWeight(index: Int): Float {
        return 1f
    }

    /**
     * titlebar标题 传空即不显示
     *
     * @return
     */
    protected abstract val title: String?

    /**
     * tab标题数组
     *
     * @return
     */
    protected abstract fun getTabNames(): Array<String>

    /**
     * tab对应的fragment
     *
     * @param position
     * @return
     */
    protected abstract fun getTabItem(position: Int): Fragment
    protected open var isAdjustMode: Boolean = true

    private fun showTabDivider(): Boolean {
        return false
    }

    /**
     * 调用此方法设置分割线样式
     *
     * @param height
     * @param color
     */
    protected fun setDivider(height: Float, color: Int) {
        val params = binding.viewDivider.layoutParams as LinearLayout.LayoutParams
        params.height = DensityUtil.dp2px(height)
        binding.viewDivider.layoutParams = params
        binding.viewDivider.setBackgroundColor(color)
        binding.viewDivider.visibility = View.VISIBLE
    }

    /**
     * 可以复用此方法自定义tab的titleview样式
     *
     * @param context
     * @param index
     * @return
     */
    protected open fun getTabTitleView(context: Context, index: Int): IPagerTitleView {
        val simplePagerTitleView: SimplePagerTitleView = ScaleTransitionPagerTitleView(context)
        simplePagerTitleView.text = mTabsName[index]
        simplePagerTitleView.textSize = 16f
        simplePagerTitleView.normalColor = -0x99999a
        simplePagerTitleView.selectedColor = -0xcdcdce
        simplePagerTitleView.setOnClickListener { binding.viewPager.currentItem = index }
        return simplePagerTitleView
    }

    /**
     * 下划线样式
     *
     * @param context
     * @return
     */
    protected open fun getTabIndicator(context: Context): IPagerIndicator {
        val indicator = LinePagerIndicator(context)
        indicator.mode = LinePagerIndicator.MODE_EXACTLY
        indicator.lineHeight = DensityUtil.dp2px(3f).toFloat()
        indicator.lineWidth = DensityUtil.dp2px(54f).toFloat()
        indicator.roundRadius = DensityUtil.dp2px(2f).toFloat()
        indicator.startInterpolator = AccelerateInterpolator()
        indicator.endInterpolator = DecelerateInterpolator(2.0f)
        indicator.setColors(-0x13e2dc)
        return indicator
    }
}